### PR TITLE
Feature/bitbox02 integration

### PR DESCRIPTION
**Pull request by Tomasvrba**
### Feature

- [x] Updated CHANGELOG.md
- [ ] Add test cases or procedure
- [x] Add PR label
- [x] No hard-coded strings

#### Please use the newest firmware `6.1.0` available in the `4.17.2` App release:
https://github.com/digitalbitbox/bitbox-wallet-app/releases
https://shiftcrypto.ch/start/

Hi @SteveMieskoski, here is an initial draft PR for the BitBox02 integration. I had never done anything in vuejs before so I'm happy to fix any issues you find. There are a few minor issues that have popped up for me after I rebased on top of `develop` that I will fix. However, I am off all next week so if you have time to review, I can address the issues when I come back on the 17th.

A few features and details to bear in mind:
- To let websites connect to the BB02, it is necessary to run the BitBoxBridge which you can find here: https://github.com/digitalbitbox/bitbox-bridge/releases
- ~I understand you currently have a BB02 beta device, this will need a firmware update which will also be released next week. We will also ship you a dev device where you will be able to load unsigned firmware like with the BB01~
- The integration should work on most major browsers on all Windows, Mac and Linux
- No u2f-hijack, popups, or external windows, the users stays on MEW the whole time
- USB communication is end-to-end encrypted with the noise protocol and the channel pairing is confirmed by the user directly on MEW
- We never redirect the user to anything hosted by us, so we can't collect metadata about the users.
- websocket closing is detected by the bridge and a callback function clears the MEW wallet interface when device is unplugged (this worked for me before rebasing on `develop` when I passed `store.dispatch` in the onClose callback, but then it broke with `clearWallet` action not recognized by `vuex`, but maybe you have a better idea what MEW callback I should pass here anyway)

- Here's a gif of what the integration looks like:
output

### CODE REVIEW

**Comment by SteveMieskoski** [comment]
@Tomasvrba Thanks for this!  We'll take a look at the issues you encountered

**Comment by Tomasvrba** [comment]
> @Tomasvrba Thanks for this! We'll take a look at the issues you encountered

Great, thanks Steve. 

One more thing I'd like to add:

Since our JS library (https://github.com/digitalbitbox/bitbox02-api-js) is  compiled with GopherJS from Go, it includes our entire Go client, not just Ethereum related functionality. So unfortunately, the final JS dependency is rather large, ~3MB which from what I saw roughly doubles the size of your `vendors` chunk. 
I tried lazy loading it with dynamic imports instead, but that didn't change anything since the MEW Vue `defaulConfig.js` bundles all node_modules into the vendors chunk. I don't want to mess with your config, but you might want to separate `node_modules/bitbox02-api` into a separate chunk.

**Comment by Tomasvrba** [comment]
@SteveMieskoski just want to ping you if you might have an estimate for when this could be reviewed and merged please. Do you think we could get to it this week? 
Happy to assist in any way I can with testing and fixing any review issues.

**Comment by Tomasvrba** [comment]
#### Please use the newest firmware `6.1.0` available in the `4.17.2` App release:
https://github.com/digitalbitbox/bitbox-wallet-app/releases
https://shiftcrypto.ch/start/

**Comment by SCBuergel** [comment]
I just tested the PR and it works like charm, kudos!

One minor remark: The app is currently cropping the data somewhere in the middle if it's too long. Mine was 68byte which is not even very long for calling some smart contract function. As a consequence, I cannot review the parameters that I pass to a function which is a pity. I suggest that you do not crop content, even if that seems lengthy. See screenshot below which is actually hiding the value of the first parameter (it's not just 0's):
IMG_20200323_233536

Let me know if I should add an issue into the bitbox firmware repo for this.

**Comment by Tomasvrba** [comment]
> I just tested the PR and it works like charm, kudos!
> 
> One minor remark: The app is currently cropping the data somewhere in the middle if it's too long. Mine was 68byte which is not even very long for calling some smart contract function. As a consequence, I cannot review the parameters that I pass to a function which is a pity. I suggest that you do not crop content, even if that seems lengthy. See screenshot below which is actually hiding the value of the first parameter (it's not just 0's):
> IMG_20200323_233536
> 
> Let me know if I should add an issue into the bitbox firmware repo for this.

Fantastic, thanks for testing @SCBuergel. 
You can definitely go ahead and open an issue here: https://github.com/digitalbitbox/bitbox02-firmware/issues
We can continue the discussion there. We are aware of the data verification issue and discussed it internally and we do have plans on how to make this better in the future. Unless the data are displayed in a human readable format in the first place though, it doesn't really matter all that much whether you can confirm that it matches or not, you still have no idea what you're actually signing (unless you are familiar with the smart contract) :/

**Comment by Tomasvrba** [comment]
@SteveMieskoski, we just released a new version with new firmware 7.0.0 which fixes some of the issues we had talked about. Since this PR is not merged yet, I will update the npm package and make a few corresponding changes on this PR asap. If you were going to merge soon-ish (today), please hold off for a day and I will make all the updates.

**Comment by NickeZ** [comment]
> I just tested the PR and it works like charm, kudos!
> 
> One minor remark: The app is currently cropping the data somewhere in the middle if it's too long. Mine was 68byte which is not even very long for calling some smart contract function. As a consequence, I cannot review the parameters that I pass to a function which is a pity. I suggest that you do not crop content, even if that seems lengthy. See screenshot below which is actually hiding the value of the first parameter (it's not just 0's):
> IMG_20200323_233536
> 
> Let me know if I should add an issue into the bitbox firmware repo for this.

Also consider that MEW will truncate long data fields. So in general it seems that the expectation from users/developers is not to be able to verify the actual hex string. Like Tom mentioned, verifying the hex string doesn't add any value because you as a user have no idea what those hex numbers mean in the general case.

It would be better if the hardware wallet identifies what smart contract it is an shows some specific information that you can verify. We already do this for ERC20 tokens, but it would be nice to support more contract types.

File: src/wallets/hardware/bitbox02/index.js [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+import { getSignTransactionObject, calculateChainIdFromV } from '../../utils';
+import errorHandler from './errorHandler';
+import store from '@/store';
+import commonGenerator from '@/helpers/commonGenerator';
+import { Misc } from '@/helpers';
+
+const NEED_PASSWORD = false;
+
+class BitBox02Wallet {
+  constructor(logout) {
```
[the remark this answers is not in the document]
Comment (kvhnuke) [reply src/wallets/hardware/bitbox02/index.js]: we'll merge it in and make this change, as it is very minor. Thanks for the PR, it is clean and nice!

**Comment by SCBuergel** [comment]
> Also consider that MEW will truncate long data fields. So in general it seems that the expectation from users/developers is not to be able to verify the actual hex string. Like Tom mentioned, verifying the hex string doesn't add any value because you as a user have no idea what those hex numbers mean in the general case.

If it truncates very long data fields (e.g. for contract deployment) that's fine. But for e.g. sending a token tx or interacting with simple functions such as Uniswap ETH-to-token swaps, I'd argue that even non-technical users (e.g. CFO) can be instructed to check the following for 4 items:
1) is the target address (this is nicely rendered in both the MEW UI as well as the Bitbox screen) correct?
2) are the first 8 characters matching `a9059cbb`?
3) ignoring the leading `0`s, are the next 20 characters matching the recipient address?
4) ignoring the leading `0`s, convert the next characters with a hex2decimal tool and is that number matching the amount to be transferred?

Without seeing that on a device screen, you have no way of verifying these. And given how complicated everyday crypto is in general I'm confident that average intelligent CFOs can be expected and instructed to handle the above.

> 
> It would be better if the hardware wallet identifies what smart contract it is an shows some specific information that you can verify. We already do this for ERC20 tokens, but it would be nice to support more contract types.

100% agree, that would be amazing but considering how after 5 years even the most advanced software wallets have only made minimal development in that direction, I don't expect this to happen on hardware wallets any time soon.

**Comment by SCBuergel** [comment]
I see this merged about a week ago but it seems not deployed yet. Any timeline on that?

**Comment by Tomasvrba** [comment]
> I see this merged about a week ago but it seems not deployed yet. Any timeline on that?

Thanks for checking in @SCBuergel, my branch was merged into the MEW repo and the MEW team took over from there getting everything ready, there is a new PR that you can follow here: https://github.com/MyEtherWallet/MyEtherWallet/pull/2158

**Pull request merged**